[Bug report] Segmentation fault when using out-of-scope variable in slices

Hi there!

I encountered a segmentation fault when using an out-of-scope variable with some slicing.

Here is an example:

`tb.sv`:
```
module testbench;

  bit [31:0] first_words [64 / 32];
  bit [31:0] second_words [32 / 32];

  bit [63:0] first;
  bit [31:0] second;

  top dut(first, second);

  always_comb begin
    for (int first_word_id = 0; first_word_id < 64 / 32; first_word_id++) begin : reset_generator
      first_words[first_word_id] = first[32*first_word_id +: 32];
    end
    for (int second_word_id = 0; second_word_id < 32 / 32; second_word_id++) begin : reset_generator
      second_words[second_word_id] = second[32*first_word_id +: 32]; // Note the usage of first_word_id here
    end
  end

endmodule
```

`top.sv`:
```
module top(first, second);
  output [63:0] first;
  wire [63:0] first;
  output [31:0] second;
  wire [31:0] second;
endmodule
```

Command: `iverilog -g2012 -o out top.sv tb.sv`

Result:
```
tb.sv:16: error: Unable to bind wire/reg/memory `first_word_id' in `testbench.$ivl_for_loop1.reset_generator'
Segmentation fault
```

The bug is still present in the most recent version of Icarus (`c9548f0`), and I'm running on Ubuntu 22.04.

Thanks!
Flavien

--- a/elab_expr.cc
+++ b/elab_expr.cc
@@ -5994,6 +5994,8 @@ NetExpr* PEIdent::elaborate_expr_net_idx_up_(Design*des, NetScope*scope,
 	    return 0;
 
       NetExpr*base = calculate_up_do_base_(des, scope, need_const);
+      if (!base)
+	    return nullptr;
 
 	// Use the part select width already calculated by test_width().
       unsigned long wid = min_width_;
@@ -6141,6 +6143,8 @@ NetExpr* PEIdent::elaborate_expr_net_idx_do_(Design*des, NetScope*scope,
 	    return 0;
 
       NetExpr*base = calculate_up_do_base_(des, scope, need_const);
+      if (!base)
+	    return nullptr;
 
 	// Use the part select width already calculated by test_width().
       unsigned long wid = min_width_;

--- a/ivtest/ivltests/partsel_invalid_idx1.v
+++ b/ivtest/ivltests/partsel_invalid_idx1.v
@@ -1,0 +1,14 @@
+// Check that a non-existent index into a vector results in an elaboration error.
+
+module test;
+
+  reg [31:0] a;
+  wire b;
+
+  assign b = a[does_not_exist]; // Error: Invalid index
+
+  initial begin
+    $display("FAILED");
+  end
+
+endmodule

--- a/ivtest/ivltests/partsel_invalid_idx2.v
+++ b/ivtest/ivltests/partsel_invalid_idx2.v
@@ -1,0 +1,14 @@
+// Check that a non-existent index into a vector results in an elaboration error.
+
+module test;
+
+  reg [31:0] a;
+  wire [1:0] b;
+
+  assign b = a[does_not_exist+:2]; // Error: Invalid base index
+
+  initial begin
+    $display("FAILED");
+  end
+
+endmodule

--- a/ivtest/ivltests/partsel_invalid_idx3.v
+++ b/ivtest/ivltests/partsel_invalid_idx3.v
@@ -1,0 +1,14 @@
+// Check that a non-existent index into a vector results in an elaboration error.
+
+module test;
+
+  reg [31:0] a;
+  wire [1:0] b;
+
+  assign b = a[does_not_exist-:2]; // Error: Invalid base index
+
+  initial begin
+    $display("FAILED");
+  end
+
+endmodule

--- a/ivtest/ivltests/partsel_invalid_idx4.v
+++ b/ivtest/ivltests/partsel_invalid_idx4.v
@@ -1,0 +1,15 @@
+// Check that a non-existent index into a parameter results in an elaboration error.
+
+module test;
+
+  parameter [31:0] P = 'h0;
+
+  wire x;
+
+  assign x = P[does_not_exist]; // Error: Invalid index
+
+  initial begin
+    $display("FAILED");
+  end
+
+endmodule

--- a/ivtest/ivltests/partsel_invalid_idx5.v
+++ b/ivtest/ivltests/partsel_invalid_idx5.v
@@ -1,0 +1,15 @@
+// Check that a non-existent index into a parameter results in an elaboration error.
+
+module test;
+
+  parameter [31:0] P = 'h0;
+
+  wire [1:0] x;
+
+  assign x = P[does_not_exist+:2]; // Error: Invalid base index
+
+  initial begin
+    $display("FAILED");
+  end
+
+endmodule

--- a/ivtest/ivltests/partsel_invalid_idx6.v
+++ b/ivtest/ivltests/partsel_invalid_idx6.v
@@ -1,0 +1,15 @@
+// Check that a non-existent index into a parameter results in an elaboration error.
+
+module test;
+
+  parameter [31:0] P = 'h0;
+
+  wire [1:0] x;
+
+  assign x = P[does_not_exist-:2]; // Error: Invalid base index
+
+  initial begin
+    $display("FAILED");
+  end
+
+endmodule

--- a/ivtest/regress-vvp.list
+++ b/ivtest/regress-vvp.list
@@ -86,6 +86,12 @@ module_ordered_list2		vvp_tests/module_ordered_list2.json
 module_port_array1		vvp_tests/module_port_array1.json
 module_port_array_init1		vvp_tests/module_port_array_init1.json
 non-polymorphic-abs		vvp_tests/non-polymorphic-abs.json
+partsel_invalid_idx1		vvp_tests/partsel_invalid_idx1.json
+partsel_invalid_idx2		vvp_tests/partsel_invalid_idx2.json
+partsel_invalid_idx3		vvp_tests/partsel_invalid_idx3.json
+partsel_invalid_idx4		vvp_tests/partsel_invalid_idx4.json
+partsel_invalid_idx5		vvp_tests/partsel_invalid_idx5.json
+partsel_invalid_idx6		vvp_tests/partsel_invalid_idx6.json
 param_test3			vvp_tests/param_test3.json
 param-width			vvp_tests/param-width.json
 param-width-vlog95		vvp_tests/param-width-vlog95.json

--- a/ivtest/vvp_tests/partsel_invalid_idx1.json
+++ b/ivtest/vvp_tests/partsel_invalid_idx1.json
@@ -1,0 +1,4 @@
+{
+    "type"          : "CE",
+    "source"        : "partsel_invalid_idx1.v"
+}

--- a/ivtest/vvp_tests/partsel_invalid_idx2.json
+++ b/ivtest/vvp_tests/partsel_invalid_idx2.json
@@ -1,0 +1,4 @@
+{
+    "type"          : "CE",
+    "source"        : "partsel_invalid_idx2.v"
+}

--- a/ivtest/vvp_tests/partsel_invalid_idx3.json
+++ b/ivtest/vvp_tests/partsel_invalid_idx3.json
@@ -1,0 +1,4 @@
+{
+    "type"          : "CE",
+    "source"        : "partsel_invalid_idx3.v"
+}

--- a/ivtest/vvp_tests/partsel_invalid_idx4.json
+++ b/ivtest/vvp_tests/partsel_invalid_idx4.json
@@ -1,0 +1,4 @@
+{
+    "type"          : "CE",
+    "source"        : "partsel_invalid_idx4.v"
+}

--- a/ivtest/vvp_tests/partsel_invalid_idx5.json
+++ b/ivtest/vvp_tests/partsel_invalid_idx5.json
@@ -1,0 +1,4 @@
+{
+    "type"          : "CE",
+    "source"        : "partsel_invalid_idx5.v"
+}

--- a/ivtest/vvp_tests/partsel_invalid_idx6.json
+++ b/ivtest/vvp_tests/partsel_invalid_idx6.json
@@ -1,0 +1,4 @@
+{
+    "type"          : "CE",
+    "source"        : "partsel_invalid_idx6.v"
+}

